# scripts and notebooks for Micro-C, ChIP-seq, ATAC-seq, and RNA-seq data analysis

### RNA-seq
<li>RNAseq_process/RNAseq_data_preprocessing_cmd.sh</li>
<p>This a example for bash command lines for RNA-seq data processing, including read mapping, read counting, generating bigwig, calculating FPKM and TPM for normalized expression index. The count matrix will be used for differential expression analysis in DESeq2 (a R package).</p>

<li>RNAseq_process/diffexp_analysis.R</li>
<p>The R script to perform differential expression analysis using DESeq2, following by gene ontology analysis and gene set enrichment ananlysis (GSEA). The input for this script is the count matrix and comparison design</p>

<li>RNAseq_process/diffexp_run.sh</li>
<p>a example bash command line to excute diffexp_analysis.R</p>

<li>RNAseq_process/comparison_design.csv</li>
<p>a example of comparison design in our study, the 1-labeled samples versus 0-labeled samples in the comparison</p>


### Micro-C
<li>MicroC/MicroC_data_preprocessing_cmd.sh</li>
<p>This a example for bash command lines for Micro-C data processing, including read mapping, read counting, generating BAM file and read pair file. The read pair file will be used in the downstream analysis for generating hic/cool file, the TAD/compartment/loop analysis will be based on hic/cool file</p>

<li>MicroC/MicroC_hic_cool_TAD_compartment_loop.sh</li>
<p>This a example for bash command lines for Micro-C data processing based on pair files generated by data preprocessing, including filtering out low-quality read pairs, generating hic/cool files, calling TAD domans and compartment, loop identification. The read pair file will be used in the downstream analysis for generating hic/cool file, the TAD/compartment/loop analysis will be based on hic/cool file.</p>

<li>MicroC/compartment.py</li>
<p>A python script to perform compartment analysis based on cool files using cooltools.</p>

### ChIP-seq and ATAC-seq
<li>ChIP_ATAC_process/chips_run.sbatch</li>
<p>an example command line for excute CHIPs pipeline for ChIP-seq or ATAC-seq data proprocessing. [CHIPs](https://f1000research.com/articles/10-517) is snakemake pipeline for chromatin data processing.</p>

<li>ChIP_ATAC_process/chips_config_for_ATAC.yaml</li>
<p>An example config file for ATAC-seq data processing, required for CHIPs pipeline</p>

<li>ChIP_ATAC_process/chips_config_for_ChIPseq.yaml</li>
<p>An example config file for ChIP-seq data processing, required for CHIPs pipeline</p>

<li>ChIP_ATAC_process/chips_metasheet_for_ATAC.csv</li>
<p>An example metasheet table for ATAC-seq data processing, required for CHIPs pipeline</p>

<li>ChIP_ATAC_process/chips_metasheet_for_ChIPseq.csv</li>
<p>An example metasheet table for ChIP-seq data processing, required for CHIPs pipeline</p>

<li>ChIP_ATAC_process/danpos_run.sh</li>
<p>An example for running [DANPOS2](https://sites.google.com/site/danposdoc/) software to call peaks for ChIP-seq and ATAC-seq data</p>

### Downstream analysis
<li>Compartment_TAD_analysis_03012024.ipynb or Compartment_TAD_analysis_03012024.py</li>
<p>The jupytor notebook or python script for Compartment and TAD downstream analysis based the outputs from proprocessing steps</p>

<li>MicroC_H2AZ_20240229.ipynb or MicroC_H2AZ_20240229.py</li>
<p>The jupytor notebook or python script for loop analysis and integration with H2AZ ChIP-seq, ATAC-seq, RNA-seq, and GWAS based the outputs from proprocessing steps</p>

<li>exp_analysis_02292024.ipynb or exp_analysis_02292024.py</li>
<p>The jupytor notebook or python script for RNA-seq data downstream analysis based the outputs from proprocessing steps</p>
