### an example bash command lines for peak calling for ChIP-seq and ATAC-seq based on plusCL_unique.sorted.bam file generated by CHIPs pipeline

bam=$1 ## path for chip bam file
Igg=$2 ## path for IgG bam file
label=$3 ## label for this sample

source /programs/biogrids.shrc
echo "++++sort and uniq bam"
samtools sort ${bam} --output-fmt BAM > ${bam}.sorted.bam
picard MarkDuplicates I=${bam}.sorted.bam O=${bam}.dedup.bam REMOVE_DUPLICATES=true ASSUME_SORTED=true VALIDATION_STRINGENCY=LENIENT METRICS_FILE=${bam}_metrices1.txt

echo "++++sort and uniq background IgG or Input"
dedup_Igg=${Igg}.dedup.bam
if [ -f "$dedup_Igg" ]; then
      echo "+++IgG dedup bam exists!"
else
      samtools sort ${Igg} --output-fmt BAM > ${Igg}.sorted.bam
      picard MarkDuplicates I=${Igg}.sorted.bam O=${dedup_Igg} REMOVE_DUPLICATES=true ASSUME_SORTED=true VALIDATION_STRINGENCY=LENIENT METRICS_FILE=${Igg}_metrices1.txt
fi

export SAMTOOLS_X=0.1.19

echo "++++Danpos to call peaks"
python.danpos2 /programs/x86_64-linux/danpos2/2.2.2/danpos.py dpeak ${bam}.dedup.bam -b ${dedup_Igg} --smooth_width 0 -o $label -c 25000000 --frsz 150 --extend 150

echo "++++get bw"
bam=`basename $bam`
bam=${bam/"/"/_}
chromsize=/lab-share/Cardio-Chen-e2/Public/rongbinzheng/Genome/mm10/mm10.chromSize
/lab-share/Cardio-Chen-e2/Public/rongbinzheng/anaconda3/bin/wigToBigWig -clip ${label}/pooled/${bam}.dedup.bgsub.Fnor.wig $chromsize ${label}/pooled/${bam}.dedup.bgsub.Fnor.wig.bw
